patch-commit fails in paths containing non-ASCII characters

### Bug report

**pnpm version:** 11.21.0 built from `main` at `ac97bc135`
**Node version:** v25.8.0
**Git version:** 2.50.1 (Apple Git-155)
**OS:** macOS 25.4.0, arm64

### Steps to reproduce

```sh
mkdir /tmp/pnpm-patch-한글
cd /tmp/pnpm-patch-한글
pnpm init
pnpm add is-positive@1.0.0
pnpm patch is-positive@1.0.0
# Edit index.js in the directory printed by pnpm.
pnpm patch-commit node_modules/.pnpm_patches/is-positive@1.0.0
```

The same failure occurs when only the custom patch edit directory contains non-ASCII characters.

### Actual behavior

`patch-commit` fails with `ERR_PNPM_INVALID_PATCH` and a `Bad diff line` error. Git quotes the non-ASCII pathname in the raw diff header, for example:

```text
diff --git a/index.js "b/.../\355\225\234\352\270\200/index.js"
```

The generated header cannot be normalized and parsed as a pnpm patch.

### Expected behavior

`patch-commit` should work independently of whether the project or edit path contains non-ASCII characters. The normalized patch header should be:

```text
diff --git a/index.js b/index.js
```

---
Written by an agent (OpenAI Codex, GPT-5).

--- a/.changeset/bright-pandas-quote.md
+++ b/.changeset/bright-pandas-quote.md
@@ -1,0 +1,7 @@
+---
+"@pnpm/patching.commands": patch
+"pnpm": patch
+"pacquet": patch
+---
+
+Fixed `pnpm patch-commit` in project and edit paths containing non-ASCII characters.

--- a/pnpm/crates/patching/src/commit.rs
+++ b/pnpm/crates/patching/src/commit.rs
@@ -149,6 +149,8 @@ pub fn diff_folders(folder_a: &Path, folder_b: &Path) -> Result<String, PatchCom
     let status = Command::new("git")
         .arg("-c")
         .arg("core.safecrlf=false")
+        .arg("-c")
+        .arg("core.quotePath=false")
         .arg("diff")
         .arg("--src-prefix=a/")
         .arg("--dst-prefix=b/")

--- a/pnpm/crates/patching/src/commit/tests.rs
+++ b/pnpm/crates/patching/src/commit/tests.rs
@@ -25,6 +25,23 @@ fn patch_commit_diff_dirs_strips_absolute_temp_paths() {
 }
 
 #[test]
+fn patch_commit_diff_dirs_supports_non_ascii_paths() {
+    let root = tempdir().expect("root dir");
+    let before = root.path().join("이전");
+    let after = root.path().join("이후");
+    fs::create_dir(&before).unwrap();
+    fs::create_dir(&after).unwrap();
+    fs::write(before.join("index.js"), "module.exports = false\n").unwrap();
+    fs::write(after.join("index.js"), "module.exports = true\n").unwrap();
+
+    let diff = diff_folders(&before, &after).expect("diff dirs");
+
+    assert!(diff.contains("diff --git a/index.js b/index.js"), "diff: {diff}");
+    assert!(!diff.contains(&before.display().to_string()), "diff: {diff}");
+    assert!(!diff.contains(&after.display().to_string()), "diff: {diff}");
+}
+
+#[test]
 fn patch_commit_diff_dirs_filters_ds_store_diffs() {
     let before = tempdir().expect("before dir");
     let after = tempdir().expect("after dir");

--- a/pnpm11/patching/commands/src/patchCommit.ts
+++ b/pnpm11/patching/commands/src/patchCommit.ts
@@ -152,7 +152,7 @@ async function diffFolders (folderA: string, folderB: string): Promise<string> {
   let stderr!: string
 
   try {
-    const result = await execa('git', ['-c', 'core.safecrlf=false', 'diff', '--src-prefix=a/', '--dst-prefix=b/', '--ignore-cr-at-eol', '--irreversible-delete', '--full-index', '--no-index', '--text', '--no-ext-diff', '--no-color', folderAN, folderBN], {
+    const result = await execa('git', ['-c', 'core.safecrlf=false', '-c', 'core.quotePath=false', 'diff', '--src-prefix=a/', '--dst-prefix=b/', '--ignore-cr-at-eol', '--irreversible-delete', '--full-index', '--no-index', '--text', '--no-ext-diff', '--no-color', folderAN, folderBN], {
       cwd: process.cwd(),
       env: {
         ...process.env,

--- a/pnpm11/patching/commands/src/patchCommit.ts
+++ b/pnpm11/patching/commands/src/patchCommit.ts
@@ -152,7 +152,7 @@ async function diffFolders (folderA: string, folderB: string): Promise<string> {
   let stderr!: string
 
   try {
-    const result = await execa('git', ['-c', 'core.safecrlf=false', '-c', 'core.quotePath=false', 'diff', '--src-prefix=a/', '--dst-prefix=b/', '--ignore-cr-at-eol', '--irreversible-delete', '--full-index', '--no-index', '--text', '--no-ext-diff', '--no-color', folderAN, folderBN], {
+    const result = await execa('git', ['-c', 'core.safecrlf=false', '-c', 'core.quotePath=false', 'diff', '--src-prefix=a/', '--dst-prefix=b/', '--ignore-cr-at-eol', '--irreversible-delete', '--full-index', '--no-index', '--text', '--no-ext-diff', '--no-color', '--', folderAN, folderBN], {
       cwd: process.cwd(),
       env: {
         ...process.env,

--- a/pnpm11/patching/commands/test/patch.test.ts
+++ b/pnpm11/patching/commands/test/patch.test.ts
@@ -387,29 +387,6 @@ describe('patch and commit', () => {
     expect(fs.readFileSync('node_modules/is-positive/index.js', 'utf8')).toContain('// test patching')
   })
 
-  test('patch and commit with a dash-prefixed relative edit dir', async () => {
-    const editDir = '-custom-edit-dir'
-
-    const output = await patch.handler({ ...defaultPatchOption, editDir }, ['is-positive@1.0.0'])
-    const patchDir = getPatchDirFromPatchOutput(output)
-
-    fs.appendFileSync(path.join(patchDir, 'index.js'), '// test patching', 'utf8')
-
-    await patchCommit.handler({
-      ...DEFAULT_OPTS,
-      cacheDir,
-      dir: process.cwd(),
-      rootProjectManifestDir: process.cwd(),
-      frozenLockfile: false,
-      fixLockfile: true,
-      storeDir,
-    }, [path.relative(process.cwd(), patchDir)])
-
-    const patchContent = fs.readFileSync('patches/is-positive@1.0.0.patch', 'utf8')
-    expect(patchContent).toContain('diff --git a/index.js b/index.js')
-    expect(fs.readFileSync('node_modules/is-positive/index.js', 'utf8')).toContain('// test patching')
-  })
-
   test('patch with relative path to custom edit dir and commit with absolute path', async () => {
     const editDir = 'custom-edit-dir'
 

--- a/pnpm11/patching/commands/test/patch.test.ts
+++ b/pnpm11/patching/commands/test/patch.test.ts
@@ -364,6 +364,29 @@ describe('patch and commit', () => {
     expect(fs.readFileSync('node_modules/is-positive/index.js', 'utf8')).toContain('// test patching')
   })
 
+  test('patch and commit with a non-ASCII edit dir', async () => {
+    const editDir = path.join(temporaryDirectory(), '한글')
+
+    const output = await patch.handler({ ...defaultPatchOption, editDir }, ['is-positive@1.0.0'])
+    const patchDir = getPatchDirFromPatchOutput(output)
+
+    fs.appendFileSync(path.join(patchDir, 'index.js'), '// test patching', 'utf8')
+
+    await patchCommit.handler({
+      ...DEFAULT_OPTS,
+      cacheDir,
+      dir: process.cwd(),
+      rootProjectManifestDir: process.cwd(),
+      frozenLockfile: false,
+      fixLockfile: true,
+      storeDir,
+    }, [patchDir])
+
+    const patchContent = fs.readFileSync('patches/is-positive@1.0.0.patch', 'utf8')
+    expect(patchContent).toContain('diff --git a/index.js b/index.js')
+    expect(fs.readFileSync('node_modules/is-positive/index.js', 'utf8')).toContain('// test patching')
+  })
+
   test('patch with relative path to custom edit dir and commit with absolute path', async () => {
     const editDir = 'custom-edit-dir'
 

--- a/pnpm11/patching/commands/test/patch.test.ts
+++ b/pnpm11/patching/commands/test/patch.test.ts
@@ -387,6 +387,29 @@ describe('patch and commit', () => {
     expect(fs.readFileSync('node_modules/is-positive/index.js', 'utf8')).toContain('// test patching')
   })
 
+  test('patch and commit with a dash-prefixed relative edit dir', async () => {
+    const editDir = '-custom-edit-dir'
+
+    const output = await patch.handler({ ...defaultPatchOption, editDir }, ['is-positive@1.0.0'])
+    const patchDir = getPatchDirFromPatchOutput(output)
+
+    fs.appendFileSync(path.join(patchDir, 'index.js'), '// test patching', 'utf8')
+
+    await patchCommit.handler({
+      ...DEFAULT_OPTS,
+      cacheDir,
+      dir: process.cwd(),
+      rootProjectManifestDir: process.cwd(),
+      frozenLockfile: false,
+      fixLockfile: true,
+      storeDir,
+    }, [path.relative(process.cwd(), patchDir)])
+
+    const patchContent = fs.readFileSync('patches/is-positive@1.0.0.patch', 'utf8')
+    expect(patchContent).toContain('diff --git a/index.js b/index.js')
+    expect(fs.readFileSync('node_modules/is-positive/index.js', 'utf8')).toContain('// test patching')
+  })
+
   test('patch with relative path to custom edit dir and commit with absolute path', async () => {
     const editDir = 'custom-edit-dir'
 

--- a/pnpm11/pnpm/test/patch/commit.ts
+++ b/pnpm11/pnpm/test/patch/commit.ts
@@ -1,0 +1,36 @@
+import fs from 'node:fs'
+import path from 'node:path'
+
+import { expect, test } from '@jest/globals'
+import { prepare } from '@pnpm/prepare'
+import { writeYamlFileSync } from 'write-yaml-file'
+
+import { execPnpmSync } from '../utils/index.js'
+
+test('patch and commit with a dash-prefixed relative edit dir', () => {
+  prepare({
+    dependencies: {
+      'is-positive': '1.0.0',
+    },
+  })
+  writeYamlFileSync('pnpm-workspace.yaml', { packages: ['.'] })
+  const cliOptions = {
+    env: {
+      pnpm_config_cache_dir: path.resolve('cache'),
+      pnpm_config_store_dir: path.resolve('store'),
+    },
+    expectSuccess: true,
+  }
+  execPnpmSync(['install'], cliOptions)
+
+  const editDir = '-custom-edit-dir'
+  execPnpmSync(['patch', 'is-positive@1.0.0', `--edit-dir=./${editDir}`], cliOptions)
+
+  fs.appendFileSync(path.join(editDir, 'index.js'), '// test patching', 'utf8')
+
+  execPnpmSync(['patch-commit', '--', editDir], cliOptions)
+
+  const patchContent = fs.readFileSync('patches/is-positive@1.0.0.patch', 'utf8')
+  expect(patchContent).toContain('diff --git a/index.js b/index.js')
+  expect(fs.readFileSync('node_modules/is-positive/index.js', 'utf8')).toContain('// test patching')
+})

